proposals/security_info_web_request.md

```diff
@@ -0,0 +1,193 @@
+   "securityInfoRawDer"
+}
+
+export interface Fingerprint {
+    /**
+     * SHA-256 hash of the certificate.
+     */ 
+    sha256: string;
+}
+
+export interface CertificateInfo {
+    fingerprint: Fingerprint;
+    rawDER: Uint8Array;
+}
+
+/**
+ * Represents the state of the network connection.
+ */
+export enum ConnectionState {
+    /**
+     * The TLS handshake failed (for example, the certificate has expired).
+     */
+    Broken = "broken",
+
+    /**
+     * The connection is not a TLS connection.
+     */
+    Insecure = "insecure",
+
+    /**
+     * The connection is a secure TLS connection.
+     */
+    Secure = "secure"
+}
+
+export interface SecurityInfo {
+    /**
+     * The array contains a single CertificateInfo object, for the server certificate, or empty in case 
+     * the connection state is "insecure" (http).
+     */
+    certificates: CertificateInfo[];
+    /**
+     * State of the connection.
+     */
+    state: ConnectionState;
+}
+
+export interface OnHeadersReceivedDetails {
+    // ... existing fields
+    /**
+     * Information about the security of the connection. This is only
+     * present if "securityInfo" is passed in the extraInfoSpec parameter
+     * and the request used a secure connection.
+     */
+    securityInfo?: SecurityInfo;
+}
+
+```
+
+### Behavior
+
+* New `onHeadersReceivedOptions` are necessary for performance reasons. They specify whether ssl data must be kept for as long as the web request is alive. Without them, unnecessary data can be kept for longer. Which is very undesirable especially in post quantum cryptography, since certificates can take a significant amount of memory space. 
+Note that Firefox has no such an option, because it only supports blocking web requests call while there’s still ssl data.
+
+* A new `securityInfo` object can be obtained in the `onHeadersReceived` event listener.
+
+* To receive this information, an extension **must** include `"securityInfo"` or `"securityInfoRawDer"` in the `extraInfoSpec` array when calling `addListener`. This opt-in design prevents performance overhead for the majority of extensions that don't need this data.
+
+* The `securityInfo` object will only be populated for requests made over a secure protocol (e.g., HTTPS, WSS) where the TLS/QUIC handshake has successfully completed or also in case of certificate errors.
+Browsers interrupt connections when there's a certificate error, unless user has explicitely allowed it in the browser UI, only in this case it is possible to have SecurityInfo with `state = "broken"`.
+
+* `certificates` - will contain only the leaf server certificate. This is done for future extensibility and Firefox API compatibility, because there they also provide a leaf if [certificateChain](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/getSecurityInfo#certificatechain) is not included in getSecurityInfo options.
```

Comment:
FYI: the majority of extensions calling `getSecurityInfo` also set the `certificateChain` option to true. Not including all certificate information may be too limiting in practice.